### PR TITLE
[REVIEW] Add cudf::test::dictionary_column_wrapper class

### PR DESCRIPTION
**Pull request by davidwendt**
This PR adds a `cudf::test::dictionary_column_wrapper` calls to the cudf gtest set of test column wrappers.

There are 2 classes defined. One that works like `fixed_width_column_wrapper` that can accept an element type and a source element type for conversion. And another that is accepts `std:string` elements:

```
Example usage:
auto col1 = cudf::test::dictionary_column_wrapper<int32_t>({2,2,3,3,0},{1,1,1,1,0});
auto col2 = cudf::test::dictionary_column_wrapper<std::string>({"b","b","cc","cc",""},{1,1,1,1,0});
```
Iterators are also supported as per the other column wrapper classes.

The new wrapper also includes a `dictionary_column_view` cast operator so that instances can be used implicitly as `column_view` parameters or passed to `cudf::dictionary::` APIs as `dictionary_column_view` objects.

A few gtests were updated to use these and I found one bug that is also fixed in this PR.

### CODE REVIEW

File: cpp/include/cudf_test/column_wrapper.hpp [hunk 1]
```diff
@@ -701,6 +702,340 @@ class strings_column_wrapper : public detail::column_wrapper {
...
+   *
+   * @note Similar to `std::vector`, this "range" constructor should be used
+   *       with parentheses `()` and not braces `{}`. The latter should only
+   *       be used for the `initializer_list` constructors.
+   *
+   * @param begin The beginning of the sequence of elements
+   * @param end The end of the sequence of elements
+   */
+  template <typename InputIterator>
+  dictionary_column_wrapper(InputIterator begin, InputIterator end) : column_wrapper{}
```
Comment (jrhemstad) [inline cpp/include/cudf_test/column_wrapper.hpp:748]: This should let you just reuse the logic already present in `fixed_width_column_wrapper`. 
```suggestion
  dictionary_column_wrapper(InputIterator begin, InputIterator end) : 
     wrapped{ encode( fixed_width_column_wrapper<KeyElementTo, SourceElementT>(begin, end) ) }
```

File: cpp/include/cudf_test/column_wrapper.hpp [hunk 2]
```diff
@@ -701,6 +702,340 @@ class strings_column_wrapper : public detail::column_wrapper {
...
+   *       with parentheses `()` and not braces `{}`. The latter should only
+   *       be used for the `initializer_list` constructors.
+   *
+   * @param begin The beginning of the sequence of elements
+   * @param end The end of the sequence of elements
+   * @param v The beginning of the sequence of validity indicators
+   */
+  template <typename InputIterator, typename ValidityIterator>
+  dictionary_column_wrapper(InputIterator begin, InputIterator end, ValidityIterator v)
+    : column_wrapper{}
```
Comment (jrhemstad) [inline cpp/include/cudf_test/column_wrapper.hpp:783]: Same thing here, just reuse the `fixed_width_column_wrapper` ctor.

File: cpp/include/cudf_test/column_wrapper.hpp [hunk 3]
```diff
@@ -701,6 +702,340 @@ class strings_column_wrapper : public detail::column_wrapper {
...
+   * // keys = {"","aaa","bbb","ccc"}, indices = {0, 1, 2, 1, 2, 3, 2}
+   * @endcode
+   *
+   * @tparam StringsIterator A `std::string` must be constructible from
+   *                         dereferencing a `StringsIterator`.
+   * @param begin The beginning of the sequence
+   * @param end The end of the sequence
+   */
+  template <typename StringsIterator>
+  dictionary_column_wrapper(StringsIterator begin, StringsIterator end) : column_wrapper{}
```
Comment (jrhemstad) [inline cpp/include/cudf_test/column_wrapper.hpp:926]: Reuse `strings_column_wrapper` ctor?

File: cpp/tests/dictionary/search_test.cpp [hunk 4]
```diff
@@ -15,83 +15,65 @@
  */
 
 #include <cudf/dictionary/detail/search.hpp>
-#include <cudf/dictionary/dictionary_column_view.hpp>
-#include <cudf/dictionary/encode.hpp>
+//#include <cudf/dictionary/dictionary_column_view.hpp>
+//#include <cudf/dictionary/encode.hpp>
```
Comment (devavret) [inline cpp/tests/dictionary/search_test.cpp:19]: remember to remove

**Review by devavret — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**